text/0000-futures.md

```diff
@@ -0,0 +1,780 @@
+
+The mechanism through which tasks get scheduled again depends on the executor
+which is driving the task.
+Possible ways of waking up a an executor include:
+- If the executor is blocked on a condition variable, the condition variable
+  needs to get notified.
+- If the executor is blocked on a system call like `select`, it might need
+  to get woken up by a syscall like `write` to a pipe.
+- If the executor's thread is parked, the wakeup call needs to unpark it.
+
+To accommodate this behavior, the `Waker` and `LocalWaker` types are
+defined through a `RawWaker` type, which is an struct that defines a dynamic
+dispatch mechanism, which consists of an raw object pointer and a virtual function
+pointer table (vtable). This mechanism allows implementers of an executor to
+customize the behavior of `Waker` and `LocalWaker` objects by providing an
+executor-specific `RawWaker`.
+
+This mechanism is chosen in favor of trait objects since it allows for more
+flexible memory management schemes. `RawWaker` can be implemented purely in
+terms of global functions and state, on top of reference counted objects, or
+in other ways.
+
+The relation between those `Waker` types is outlined in the following definitions:
+
+```rust
+/// A `RawWaker` allows the implementor of a task executor to customize the
+/// behavior of `LocalWaker`s and `Waker`s.
+/// It consists of a data pointer and a virtual function pointer table (vtable) that
+/// customizes the behavior of the `RawWaker`.
+#[derive(PartialEq)]
+pub struct RawWaker {
+    /// A data pointer, which can be used to store arbitrary data as required
+    /// by the executor. This could be e.g. a type-erased pointer to an `Arc`
+    /// that is associated with the task.
+    /// The value of this field gets passed to all functions that are part of
+    /// the vtable as first parameter.
+    pub data: *const (),
+    /// Virtual function pointer table that customizes the behavior of this waker.
+    pub vtable: &'static RawWakerVTable,
+}
+
+/// A virtual function pointer table (vtable) that allows to customize the
+/// behavior of a `RawWaker`.
+/// The pointers which get passed to all functions inside the vtable are the
+/// values of the `data` field.
+#[derive(PartialEq, Copy, Clone)]
+pub struct RawWakerVTable {
+    /// This function will be called when the `RawWaker` gets cloned, e.g. when
+    /// the `Waker` or `LocalWaker` in which the `RawWaker` is stored gets cloned.
+    ///
+    /// The implementation of this function must retain all resources that are
+    /// required for this additional instance of a `RawWaker` and associated
+    /// task. The implementation must return a valid `RawWaker` that behaves
+    /// equivalent to the `RawWaker` that got cloned. E.g. cloning a `RawWaker`
+    /// that implemented a thread-safe wakeup for use in `Waker` must return
+    /// a `RawWaker` that implements the same wakeup behavior.
+    pub clone: unsafe fn(*const ()) -> RawWaker,
```

Comment:
This returns a `RawWaker`, but the `Clone` implementation expects data `*const ()`, see https://github.com/rust-lang/rfcs/pull/2592/files#diff-bf8a6ad34e0f1cb875733c92867bb0f9R325.

Reply:
In case you are interested: It originally returned only `*const ()` - which means it was only intended to change the instance data, but not the vtable. While porting the current code I figured out that `LocalWakerRef` wasn't implementable in a convenient fashion with that. Since for this we want to get a different vtable too (from a Fake instance to real one). That's why the function now returns `RawWaker`.

Reply:
Yup, makes sense.